### PR TITLE
cmake builds libaegis

### PR DESCRIPTION
**Pull request by victorstewart**

### CODE REVIEW

**Comment by jedisct1** [comment]
Thank you!

The reference to an external repository may be an issue for deployment and packaging.

It may be better to allow users to specify the `libaegis` directory.

**Comment by victorstewart** [comment]
oh right that makes sense— i had assumed you meant replace the existing code with an automated build.

so another path allowing to specify the include directory and another for the library file?

**Comment by jedisct1** [comment]
1) The library is already compiled and installed on the system, so we can try to link it and find it in the default header paths (or via `pkg-config`), and if they are not found, that can be overridden in the command line.
2) The library is not installed, so we can specify the path to the source code, and it will compile it and link it.

**Comment by victorstewart** [comment]
for 2, do you mean a local path on disk? or an internet path?

if on disk, i'm thinking this series might make most sense

```
IF (AEGIS_SOURCE)
   // they provided for example AEGIS_SOURCE=/root/libaegis so we compile it and link it
ELSE()
   FIND_PACKAGE(aegis)
   IF (AEGIS_FOUND)
      // we found libaegis installed so we link the library and include the headers
   ELSE()
      // we download libaegis from GitHub then compile and link it
   ENDIF()
ENDIF()
```

i think the idea of finding the library file via FIND_PACKAGE but then the include directory being able to not be in the same directory, but able to be independently specified... just feels weird to me?

if we really want to be able to specify where to look, i think it could make sense to provide AEGIS_DIR and let FIND_PACKAGE look there?

**Comment by jedisct1** [comment]
Yes, a local path on the disk.

When installed on the system, the library can be in something like `/usr/local/lib` and the include file in `/usr/local/include`, so different directories. But if we generate a config for `pkg-config`, these paths will be found automatically.

**Comment by victorstewart** [comment]
what i meant to say was not in the same parent directory. if "make install" generated a pkg-config and installed the library file into /usr/local/lib and the headers into /usr/local/include, then i don't understand why the include path would need to be explicitly specified because FIND_PACKAGE would find both? that was my question. i don't see how that branch would ever be taken.

**Comment by victorstewart** [comment]
but assuming that can occur, the below covers all options. otherwise i think the entire `IF (NOT EXISTS "${AEGIS_INCLUDE_DIRS}/aegis.h")` could be culled?

(and the FIND_PACKAGE(aegis) command is spurious sometimes but this was the only way that came to mind to unite the two build branches, unless we added yet another external flag, which i didn't like the idea of)

```
IF (WITH_AEGIS)

    MESSAGE(STATUS "  Enabling AEGIS support")
    SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DPTLS_HAVE_AEGIS=1")

    FIND_PACKAGE(aegis)

    IF (AEGIS_SOURCE OR NOT AEGIS_FOUND)

        include(ExternalProject)
        SET(AEGIS_AUTOBUILD TRUE)

        IF (AEGIS_SOURCE)

            MESSAGE(STATUS "building libaegis from ${AEGIS_SOURCE}")
            ExternalProject_Add(libaegis
                SOURCE_DIR ${AEGIS_SOURCE}
                BUILD_COMMAND make -j4
                BINARY_DIR ${AEGIS_SOURCE}/build
            )

        ELSE()

            MESSAGE(STATUS "downloading libaegis 0.1.1 from GitHub")
            set(AEGIS_SOURCE libaegis)
            ExternalProject_Add(libaegis

                GIT_REPOSITORY https://github.com/jedisct1/libaegis.git
                GIT_TAG 0.1.1
                GIT_SHALLOW ON
                SOURCE_DIR ${AEGIS_SOURCE}
                BUILD_COMMAND make -j4
                BINARY_DIR ${AEGIS_SOURCE}/build
            )

        ENDIF()

        SET(AEGIS_LIBRARIES ${AEGIS_SOURCE}/build/libaegis.a)
        INCLUDE_DIRECTORIES(${AEGIS_SOURCE}/src/include)

    ELSE()

        MESSAGE(STATUS "found libaegis in ${AEGIS_DIR})")

        # AEGIS_LIBRARIES already set by FIND_PACKAGE
        # AEGIS_INCLUDE_DIRS already set by FIND_PACKAGE
        
        IF (NOT EXISTS "${AEGIS_INCLUDE_DIRS}/aegis.h")
            IF (AEGIS_INCLUDE_DIR AND EXISTS "${AEGIS_INCLUDE_DIR}/aegis.h")
                INCLUDE_DIRECTORIES(${AEGIS_INCLUDE_DIR})
                MESSAGE(STATUS "aegis.h not found in ${AEGIS_DIR}) so using your override in ${AEGIS_INCLUDE_DIR}")
            ELSE()
                IF (AEGIS_INCLUDE_DIR)
                    MESSAGE(FATAL_ERROR "aegis.h not found in ${AEGIS_INCLUDE_DIRS} or ${AEGIS_INCLUDE_DIR}")
                ELSE()
                    MESSAGE(FATAL_ERROR "aegis.h not found in ${AEGIS_INCLUDE_DIRS}. define AEGIS_INCLUDE_DIR accordingly")
                ENDIF()
            ENDIF()
        ELSE()
            INCLUDE_DIRECTORIES(${AEGIS_INCLUDE_DIRS})
        ENDIF()

    ENDIF()

ENDIF()
```

**Comment by jedisct1** [comment]
> what i meant to say was not in the same parent directory. if "make install" generated a pkg-config and installed the library file into /usr/local/lib and the headers into /usr/local/include, then i don't understand why the include path would need to be explicitly specified because FIND_PACKAGE would find both? that was my question. i don't see how that branch would ever be taken.

Sorry for the confusion. The parent directory can indeed be assumed to be the same.

`pkg-config` doesn't work well on Windows, so we need an alternative way to specify the prefix. But if `FIND_PACKAGE` can find also packages in manually specified prefixes, this is good! You know `cmake` far better than I do.

**Comment by jedisct1** [comment]
> `BUILD_COMMAND make -j4`

Is that necessary? Can't the same command than the one used to build `picotls` be used, so that it also works with Xcode, ninja and Visual Studio?

**Comment by jedisct1** [comment]
With `FIND_PACKAGES`, the following variables get set:

```
-- aegis_CMAKE_DIR=/tmp/op/share/cmake/aegis
-- aegis_CONFIG=/tmp/op/share/cmake/aegis/aegis-config.cmake
-- aegis_CONSIDERED_CONFIGS=/tmp/op/share/cmake/aegis/aegis-config.cmake
-- aegis_CONSIDERED_VERSIONS=unknown
-- aegis_DIR=/tmp/op/share/cmake/aegis
-- aegis_FOUND=1
-- aegis_LIBRARIES=aegis::aegis
-- aegis_VERSION_COUNT=0
-- aegis_VERSION_MAJOR=0
-- aegis_VERSION_MINOR=0
-- aegis_VERSION_PATCH=0
-- aegis_VERSION_TWEAK=0
```

The `include` directory is not present. Also, how to set the version?

**Comment by victorstewart** [comment]
see the latest code i just pushed. tested all 3 branches. we should be prefect here.

the cmake package config file you created here https://github.com/jedisct1/libaegis/blob/b5f43197e0d56b7dc26e35edfd9651661e058000/CMakeLists.txt#L52C1-L52C30 carries the information about where the header files are. so when you add that aegis:aegis target to another object with like TARGET_LINK_LIBRARIES, the header files get pulled into visibly as well.

**Comment by victorstewart** [comment]
so aegis can be turned on with either -DWITH_AEGIS=ON (which first chooses a found installed package, otherwise downloads version 0.1.2 from GitHub) or -DWITH_AEGIS=/root/libaegis-0.1.2 which builds that code (out of tree).

the only possibly questionable thing is currently your compile commands auto installs the library... we might not want that in any of these 3 branches? like spooky action at a distance installing things on the user's system without warning or option.

**Comment by victorstewart** [comment]
i rebased my build PR on your latest code.

are you happy with it?

File: lib/openssl.c [hunk 1]
```diff
@@ -2290,4 +2290,4 @@ ptls_hpke_cipher_suite_t *ptls_openssl_hpke_cipher_suites[] = {&ptls_openssl_hpk
                                                                &ptls_openssl_hpke_chacha20poly1305sha256,
 #endif
                                                                &ptls_openssl_hpke_aes128gcmsha512, /* likely only for tests */
-                                                               NULL};
+                                                               NULL};
```
Comment (victorstewart) [inline lib/openssl.c:2293]: it created this annoying little thing when i was merging the conflicts 🙄

**Pull request closed without merging**